prov/verbs_ud/src/verbs_ud.h

```diff
@@ -0,0 +1,738 @@
+
+	uint16_t	lid; /* LRH */
+	uint16_t	pkey; /* BTH */
+	uint16_t	service; /* for NS src addr, 0 means any */
+
+	uint8_t 	sl; /* LRH */
+	uint8_t		padding; /* forced padding to 256 bits (32 byte) */
+}; /* 256 bits */
+
+struct ofi_verbs_ud_fabric {
+	struct util_fabric	util_fabric;
+	struct util_ns		name_server;
+};
+
+struct ofi_verbs_ud_domain {
+	struct util_domain	util_domain;
+	struct util_av		*av;
+    	struct fi_info		*info;
+	struct ibv_comp_channel	*ibv_comp_channel;
+	struct ibv_context	*ibv_device_ctx;
+	int			port_number;
+	struct ibv_port_attr	*port_attr;
+	struct ibv_pd		*ibv_pd;
+	/* NULL if gl_data::separate_send_recv_cq = 1 */
+	struct ibv_cq		*ibv_shared_cq;
+};
+
+struct ofi_verbs_ud_ep {
+	struct util_ep			util_ep;
+	struct ibv_qp			*ibv_qp;
+	struct fi_info			*info;
+	struct ofi_verbs_ud_av		*av;
+	struct ofi_verbs_ud_domain	*domain;
+	struct ibv_cq			*ibv_shared_cq;
+	struct ibv_cq			*ibv_send_cq;
+    	struct ibv_cq			*ibv_recv_cq;
+	struct ofi_verbs_ud_buf_pool	grh_pool;
+	struct ofi_ib_ud_ep_name	ep_name;
+	int				service;
+	int				ep_flags;
+	struct ibv_wc			*multiple_wcs;
+};
+
+struct ofi_verbs_ud_cq {
+	struct util_cq	util_cq;
+	struct ibv_cq	*ibv_cq;
+};
+
+struct ofi_verbs_ud_mem_desc {
+	struct fid_mr			mr_fid;
+	struct ibv_mr			*mr;
+	struct ofi_verbs_ud_domain	*domain;
+};
+
+struct ofi_verbs_ud_av {
+	struct util_av	util_av;
+	RbtHandle	addr_map;
+};
+
+struct ofi_verbs_ud_av_entry {
+	struct ofi_ib_ud_ep_name	*addr;
+	struct ibv_ah			*ah;
+};
+
+extern struct util_prov ofi_verbs_ud_util_prov;
+extern struct fi_provider ofi_verbs_ud_prov;
+
+extern const char ofi_verbs_ud_prov_name[];
+extern const char ofi_verbs_ud_fab_name[];
+extern const char ofi_verbs_ud_domain_name[];
+
+#define VERBS_UD_WARN(subsys, ...)					\
+	FI_WARN(&ofi_verbs_ud_prov, subsys, __VA_ARGS__)
+#define VERBS_UD_DBG(subsys, ...)					\
+	FI_DBG(&ofi_verbs_ud_prov, subsys, __VA_ARGS__)
+#define VERBS_UD_INFO(subsys, ...)					\
+	FI_INFO(&ofi_verbs_ud_prov, subsys, __VA_ARGS__)
+#define VERBS_UD_INFO_ERRNO(subsys, fn, errno)				\
+	VERBS_UD_INFO(subsys, fn ": %s(%d)\n", strerror(errno), errno)
+
+/* Verbs UD Pool functionality */
+struct ofi_verbs_ud_wr_entry_hdr {
+	struct dlist_entry	entry;
+	void			*desc;
+	struct ofi_verbs_ud_ep	*ep;
+	void			*context;
+	uint64_t		flags;
+};
+
+struct ofi_verbs_ud_pool_attr {
+	size_t	count;
+	size_t	size;
+	void	*pool_ctx;
+
+	ofi_verbs_ud_pool_entry_cancel_hndlr cancel_hndlr;
+
+	util_buf_region_alloc_hndlr	alloc_hndlr;
+	util_buf_region_free_hndlr	free_hndlr;
+};
+
+
+/*
+ * The Global Routing Header (GRH) of the incoming message
+ * will be placed in the first 40 bytes of the buffer(s)
+ * in the scatter list.
+ * @note If no GRH is present in the incoming message,
+ *       then the first bytes will be undefined.
+ * This means that in all cases, the actual data of the
+ * incoming message will start at an offset of 40 bytes
+ * into the buffer(s) in the scatter list.
+ */
```

Comment:
If you don't actually need the grh, you can use a single buffer for all receives to receive the grh in.